src/audio/aec/aec.c

```diff
@@ -0,0 +1,543 @@
+	int32_t out;
+	int32_t ref;
+	int32_t *x;
+	int32_t *y;
+	int32_t *r32;
+	int16_t *r16;
+	int in_idx;
+	int ref_idx;
+	int ref_nch;
+	int i;
+	int j;
+	int nch = sources[0]->channels;
+	int out_idx = 0;
+
+	if (cd->ref_active)
+		ref_nch = sources[1]->channels;
+
+	/* Input from selected input and reference channels, same output to every channel */
+	in_idx = cd->source_channel;
+	ref_idx = cd->reference_channel;
+	for (i = 0; i < frames; i++) {
+		x = audio_stream_read_frag_s32(sources[0], in_idx);
+		mix = *x << 8;
+		in_idx += nch;
+		if (cd->ref_active) {
+			if (cd->ref_32bits) {
+				r32 = audio_stream_read_frag_s32(sources[1], ref_idx);
+				ref = *r32;
+			} else {
+				r16 = audio_stream_read_frag_s16(sources[1], ref_idx);
+				ref = *r16;
+			}
+
+			ref_idx += ref_nch;
+			mix = mix - (ref << cd->ref_shift);
+		}
+
+		out = sat_int24(mix >> 8);
+		for (j = 0; j < nch; j++) {
+			y = audio_stream_write_frag_s32(sink, out_idx);
+			*y = out;
+			out_idx++;
+		}
+	}
+}
+
+#endif
+
+#if CONFIG_FORMAT_S32LE
+
+static void aec_s32_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames)
+
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int64_t mix;
+	int32_t out;
+	int32_t ref;
+	int32_t *x;
+	int32_t *y;
+	int32_t *r32;
+	int16_t *r16;
+	int in_idx;
+	int ref_idx;
+	int ref_nch;
+	int i;
+	int j;
+	int nch = sources[0]->channels;
+	int out_idx = 0;
+
+	if (cd->ref_active)
+		ref_nch = sources[1]->channels;
+
+	/* Input from selected input and reference channels, same output to every channel */
+	in_idx = cd->source_channel;
+	ref_idx = cd->reference_channel;
+	for (i = 0; i < frames; i++) {
+		x = audio_stream_read_frag_s32(sources[0], in_idx);
+		mix = *x;
+		in_idx += nch;
+		if (cd->ref_active) {
+			if (cd->ref_32bits) {
+				r32 = audio_stream_read_frag_s32(sources[1], ref_idx);
+				ref = *r32;
+			} else {
+				r16 = audio_stream_read_frag_s16(sources[1], ref_idx);
+				ref = *r16;
+			}
+
+			ref_idx += ref_nch;
+			mix = mix - (ref << cd->ref_shift);
+		}
+
+		out = sat_int32(mix);
+		for (j = 0; j < nch; j++) {
+			y = audio_stream_write_frag_s32(sink, out_idx);
+			*y = out;
+			out_idx++;
+		}
+	}
+}
+
+#endif
+
+/* Processing functions table */
+const struct aec_func_map aec_fnmap[] = {
+#if CONFIG_FORMAT_S16LE
+	{ SOF_IPC_FRAME_S16_LE, aec_s16_default },
+#endif
+#if CONFIG_FORMAT_S24LE
+	{ SOF_IPC_FRAME_S24_4LE, aec_s24_default },
+#endif
+#if CONFIG_FORMAT_S32LE
+	{ SOF_IPC_FRAME_S32_LE, aec_s32_default },
+#endif
+};
+
+const size_t aec_fncount = ARRAY_SIZE(aec_fnmap);
+
+/*
+ * End of AEC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *aec_new(const struct comp_driver *drv, struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *aec;
+	struct sof_ipc_comp_process *ipc_aec = (struct sof_ipc_comp_process *)comp;
+	int ret;
+
+	comp_cl_info(&comp_aec, "aec_new()");
```

Comment:
debug please

Reply:
Yep, I need to learn to use the trace filters for sof-logger.